src/crypto/bort.cpp

```diff
@@ -0,0 +1,145 @@
+#include "bort.hpp"
+#include "crypto/sha256.h"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+
+void clamp_curve25519(unsigned char *key)
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char *K,
+    const uint256 &dhsecret,
+    const uint256 &in_epk,
+    const uint256 &pk_enc,
+    unsigned char in_nonce
+   )
+{
+    unsigned char distinguisher = 0xC2;
+
+    CSHA256 hasher;
+    hasher.Write(&distinguisher, 1);
+    hasher.Write(dhsecret.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(in_epk.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(pk_enc.begin(), crypto_scalarmult_BYTES);
+    hasher.Write(&in_nonce, 1);
+    hasher.Finalize(K);
+}
+
+template<size_t MLEN>
+Bort<MLEN>::Bort() {
+    // All of this code assumes crypto_scalarmult_BYTES is 32
+    // There's no reason that will _ever_ change, but for
+    // completeness purposes, let's check anyway.
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_BYTES);
+    BOOST_STATIC_ASSERT(BORT_AUTH_BYTES == crypto_aead_chacha20poly1305_ABYTES);
+
+    // Create the ephemeral keypair
+    randombytes_buf(esk.begin(), crypto_scalarmult_BYTES);
+    epk = generate_pubkey(esk);
+}
+
+template<size_t MLEN>
+Bort<MLEN>::~Bort() {
+    sodium_memzero(epk.begin(), crypto_scalarmult_BYTES);
+    sodium_memzero(esk.begin(), crypto_scalarmult_BYTES);
+}
+
+template<size_t MLEN>
+typename Bort<MLEN>::Ciphertext Bort<MLEN>::encrypt(const uint256 &pk_enc,
+                                                    const Bort<MLEN>::Plaintext &message,
+                                                    unsigned char in_nonce
+                                                   )
+{
+    uint256 dhsecret;
+
+    if (crypto_scalarmult(dhsecret.begin(), esk.begin(), pk_enc.begin()) != 0) {
```

Comment:
What do you prefer we do instead here? Triggering an error here appears to be how the libsodium documentation recommended downstream users behave.